homeassistant/components/scheduler/__init__.py

```diff
@@ -0,0 +1,432 @@
+        def store_entity_if_in_context(
+            entity_id: str, old_state: "State", new_state: "State"
+        ) -> None:
+            """Save prior states of an entity if it was triggered by this schedule."""
+            if new_state.context == self._activation_context:
+                self._entity_states[entity_id] = old_state
+
+        self._async_state_listener = async_track_state_change(
+            self._hass, MATCH_ALL, store_entity_if_in_context
+        )
+
+        _LOGGER.info("Scheduler activated scene: %s", self.entity_id)
+
+    @callback
+    def async_undo_scene(self) -> None:
+        """Restore the entities touched by the schedule."""
+        if not self._entity_states:
+            return
+
+        for entity_id, state in self._entity_states.items():
+            self._hass.states.async_set(
+                entity_id,
+                state.state,
+                attributes=state.attributes,
+                context=self._activation_context,
+            )
+
+        _LOGGER.info("Scheduler deactivated scene: %s", self.entity_id)
+
+
+class Scheduler:
+    """A class to manage scene schedules."""
+
+    def __init__(self, hass: "HomeAssistant") -> None:
+        """Initialize."""
+        self._async_activation_listeners: Dict[str, Callable] = {}
+        self._async_deactivation_listeners: Dict[str, str] = {}
+        self._hass: "HomeAssistant" = hass
+        self._latest_instances: List[ScheduleInstance] = []
+        self.schedules: Dict[str, dict] = {}
+
+    @callback
+    def as_dict(self, schedule_id: str) -> dict:
+        """Return a schedule in dict form."""
+        schedule = self.schedules[schedule_id]
+        return {
+            CONF_SCHEDULE_ID: schedule_id,
+            CONF_ENTITY_ID: schedule[CONF_ENTITY_ID],
+            CONF_START_DATETIME: schedule[CONF_START_DATETIME].isoformat(),
+            CONF_DURATION: schedule[CONF_DURATION],
+            CONF_RRULE: str(schedule[CONF_RRULE]),
+            CONF_ACTIVATION_CONTEXT_ID: schedule[CONF_ACTIVATION_CONTEXT].id,
+        }
+
+    @callback
+    def async_create(
+        self,
+        entity_id: str,
+        start_datetime: datetime,
+        *,
+        duration: Optional[int] = None,
+        rrule_str: Optional[str] = None,
+        activation_context_id: Optional[str] = None,
+    ) -> dict:
+        """Create a schedule.
+
+        The rrule_str parameter must be an RFC5545-compliant string.
+        The duration parameter is a number of seconds a schedule instance should last.
+        """
+        if start_datetime < datetime.now():
+            raise ValueError("Dates in the past are not allowed")
+
+        schedule_id = uuid4().hex
+        schedule = {
+            CONF_ENTITY_ID: entity_id,
+            CONF_START_DATETIME: start_datetime,
+            CONF_DURATION: duration,
+            CONF_RRULE: rrulestr(rrule_str) if rrule_str else None,
+            CONF_ACTIVATION_CONTEXT: Context(id=activation_context_id)
```

Comment:
I'm confused why you would store a context ? The context is created WHEN you activate it. It's unique for each activation.